docs/transformations/CustomTransformation.md

```diff
@@ -0,0 +1,212 @@
+nav_order: 15
+permalink: /custom-transformation
+---
+
+# Custom transformation
+{: .no_toc }
+Users can follow this guide to apply custom transformations at the row level to all columns in their schema, including primary key columns. This guide can be used to specify custom transformation logic for existing columns, to populate new columns in Spanner that didn't exist in the source, or to filter a row during migration. For ease of use, we have implemented the [CustomTransformationFetcher](https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/main/v2/spanner-custom-shard/src/main/java/com/custom/CustomTransformationFetcher.java) class in `v2/spanner-custom-shard`, which can be updated according to the user's logic.
+
+<details open markdown="block">
+  <summary>
+    Table of contents
+  </summary>
+  {: .text-delta }
+1. TOC
+{:toc}
+</details>
+
+## Custom transformation workflow
+
+![](https://services.google.com/fh/files/misc/transformation-workflow.png)
+
+Points to note regarding the workflow:
+- The workflow will execute once per event and will also be triggered during retries.
+- Not all column values need to be returned; only the returned columns will be updated. The rest will be migrated as they are from the source.
+- Users can add new columns in the response that are not present in the source but exist in Spanner. These column values will be written to Spanner, and the data types of the returned values must be compatible with the Spanner schema.
+
+## Methods in CustomTransformationFetcher
+
+- **init()** - This is an initialization method that will be called once during the pipeline setup. It is used to initialize the custom jar with custom parameters.  
+- **toSpannerRow()** - This method applies custom transformations to the incoming source record and is expected to return a subset of spanner row.
+- **toSourceRow()** - This method applies custom transformations to the incoming spanner record and is expected to return a subset of source row.
+
+## Parameter details
+
+### Forward migration
+
+#### Request
+[MigrationTransformationRequest](https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/main/v2/spanner-migrations-sdk/src/main/java/com/google/cloud/teleport/v2/spanner/utils/MigrationTransformationRequest.java) contains the following information - 
+- tableName - Name of the source table to which the event belongs to.
+- shardId - Logical shard id of the record.
+- eventType - The event type can either be INSERT, UPDATE-INSERT, UPDATE, UPDATE_DELETE or DELETE. Please refer to the [datastream documentation](https://cloud.google.com/datastream/docs/events-and-streams) for more details.
+- requestRow - It is a map of type `Map<java.lang.String, java.lang.Object>` where key is the source column name and value is source column value.
+
+The following table outlines the Java object type that will be sent in the request to the `toSpannerRow` for each **MySQL** datatype:
+
+| MYSQL datatype | Java object type    |
```

Comment:
Thanks, this looks much better.

nit: for datetime and timestamp types, instead of eg, maybe we should say format to indicate utc is always the format.